### PR TITLE
Add a reason code to error messages on both sides.

### PR DESCRIPTION
**Pull request by lynxis**

### CODE REVIEW

File: client/l2tp_client.c [hunk 1]
```diff
@@ -839,10 +861,11 @@ int context_session_set_mtu(l2tp_context *ctx)
   return 0;
 }
 
-void context_close_tunnel(l2tp_context *ctx)
+void context_close_tunnel(l2tp_context *ctx, uint8_t reason)
 {
+  reason |= ERROR_DIRECTION_CLIENT;
```
Comment (kostko) [inline client/l2tp_client.c]: Add an empty line after this line.

**Comment by kostko** [comment]
Nice, this feature should help with debugging link problems! Looks good on first glance and it looks like it should also gracefully handle combinations of old and new brokers/clients (just without the error code). Have you done any quick tests with different broker/client version combinations, just to be sure?

**Comment by lynxis** [comment]
> Have you done any quick tests with different broker/client version combinations, just to be sure?

honestly. not really. We need proper auto testing. Maybe I can spend next week a evening for testing.
Tested with new tunneldigger server <> new clients.

**Comment by mitar** [comment]
Yea, it would be great if we could have some testing suite and CI server. @kostko, didn't you say that CircleCI gives you access to network stack?

**Comment by kostko** [comment]
I've read in their documentation that they support running Docker containers. So it would be possible to do this inside a docker container.

**Comment by lynxis** [comment]
@kostko @mitar I've started a test script. See https://github.com/lynxis/tunneldigger-testing it's using lxc containers. I'll setup a jenkins later and connect it to my github tunneldigger repository for testing it with pull requests.

**Comment by kostko** [comment]
This looks great! Can we move it under wlanslovenija organization and give you full permissions there?

**Comment by lynxis** [comment]
it's near ready. the jenkins' gitrepo must be used, multiple revision should be tested, clean up of old containers.

**Comment by lynxis** [comment]
so testing is ready. see PR in lynxis/tunneldigger repo.
like https://github.com/lynxis/tunneldigger/pull/7
My test run does:
# client <> server
newcommit<>newcommit
newcommit<>7cbe5d60c1b72415e15e573e0a9189f47a3f2094
newcommit<>c638231efca6b3a6e1c675ac0834a3e851ad1bdc
7cbe5d60c1b72415e15e573e0a9189f47a3f2094<>newcommit
c638231efca6b3a6e1c675ac0834a3e851ad1bdc<>newcommit

The 2 git revs were taken by random. maybe you have a better idea, which versions to take?

**Comment by lynxis** [comment]
Also we have to add my jenkins to the hooks of this repository.

**Comment by mitar** [comment]
Are you using jenkins or Travis CI? The idea was that we should try to use one of existing providers, like Travis CI or Circle CI?

**Comment by lynxis** [comment]
i'm using jenkins. If you like I can give you two accounts to the jenkins.
@mitar if you find a provider who supports lxc and network setups.

**Comment by lynxis** [comment]
I though testing is missing. Now we can test this revision/pr against multiple revision.
@mitar @kostko I would like finish this pull request. so what do we need, so we can merge this?
If you tell me against what revision I should test, I will do a test run against that.

**Comment by mitar** [comment]
So we should have some tests for comparing various versions, if they are still able to talk to each other after this change. So is server is old and client is new, and if client is new and server is old.

**Comment by lynxis** [comment]
@mitar I implemented that part. Against what revision should I test?
Atm the testing takes 9min30seconds including 5 test run. I'm testing against two "stable" revisions. A test run is one server+client test
5 test runs = new+new, new+oldA, new+oldB, oldA+new, oldB+new = 9min30sec

**Comment by mitar** [comment]
Test against the latest `master` and your branch.

I will leave the rest to @kostko to check.

But we should really get this testing setup for all pull requests here.

**Comment by kostko** [comment]
> Test against the latest master and your branch.

Yes.

> But we should really get this testing setup for all pull requests here.

I agree, but this should go into the next pull request.

**Comment by lynxis** [comment]
To get pull-requests etc. the jenkins needs 2 webhooks + 1 permission:
- webhook->plain: https://jenkins.fe80.eu/ghprbhook/ with notification pull-request + issue comment
- webhook->service: jenkins (github plugin) - weburl https://jenkins.fe80.eu/github-webhook/
-grant permission **repo:status** to user @fe80-bot

**Comment by lynxis** [comment]
https://github.com/lynxis/tunneldigger/pull/13

**Comment by kostko** [comment]
So you tested current master against the changes introduced by this PR?

**Comment by lynxis** [comment]
[consoleText.txt](https://github.com/wlanslovenija/tunneldigger/files/43394/consoleText.txt)
attached the test report. I tested against 3 versions OLD_REV="c638231efca6b3a6e1c675ac0834a3e851ad1bdc 7cbe5d60c1b72415e15e573e0a9189f47a3f2094 4e4f13cdc630c46909d47441093a5bdaffa0d67f"

testing against the "master" symbol is unlikely, because it won't be reproducible, so I added the sha of the master 4e4f13cdc is the actual master.

**Comment by kostko** [comment]
Ok, great, this looks good now. You can create another pull request for integrating the test environment into the current tunneldigger. Would it be possible to run these tests on Travis CI?

**Pull request merged**